README.md
# [![Linux for PHP Banner](https://linuxforphp.net/download_file/force/30/0)](https://linuxforphp.net/)

https://linuxforphp.net

# Linux for PHP CI Kit for GitLab using Docker

[![Build Status](https://travis-ci.org/linuxforphp/docker-gitlab-ci.svg?branch=master)](https://travis-ci.org/linuxforphp/docker-gitlab-ci)

This repository contains a set of utilities for running PHP tests via [GitLab CI](https://about.gitlab.com/gitlab-ci/).

These tools provide:

1. [GitLab Runner](https://gitlab.com/gitlab-org/gitlab-ci-multi-runner) deployment script for Ubuntu on VM or metal server. It is not recommended to install runner right on the production system.

2. Set of Docker images for PHP 5.6 - 7.3 based on official Linux for PHP Docker images, with all essential PHP modules. The images also include Node.js.

The goal of these tools is to automate as much as possible the routine work that is related to configuring the GitLab Runner, so you can concentrate on writing tests for your code.
Also these tools are trying to be resource savvy, since in most cases, huge in-RAM caches are not needed for only running unit tests with some fixtures. So, very small VMs can be suitable for running tests.

## Contents of repository
 - [gitlab-runner bootstrap script](https://github.com/linuxforphp/docker-gitlab-ci/tree/master/gitlab-runner-vm) for deploying GitLab Runner.
 - [PHP Docker images](https://github.com/linuxforphp/docker-gitlab-ci/tree/master/php) - based on official Linux for PHP images, including all major modules and extensions. Also, PHP 5.6 is available.
 - [Example projects](https://github.com/linuxforphp/docker-gitlab-ci/tree/master/examples) - Example PHP test script.

## Quick start

1. Install [GitLab](https://about.gitlab.com/),
2. Get a server (VM or metal) with minimal Ubuntu-14.04 installed. It will be used for the runner.
3. Login as root to a server and execute `curl -S https://raw.githubusercontent.com/linuxforphp/docker-gitlab-ci/master/gitlab-runner-vm/bootstrap.sh | bash` and answer the questions. This script will install Docker, the Gitlab Runner, and configure the Runner for using these Docker images.

Runner is limited to `asclinux/linuxforphp-8.1-ultimate:*` images for main container (where your repository is cloned) and any service images `*/*` (secondary containers that are spun up for services, like MySQL, Redis, etc).

If you want to use the server for also running other images (`ruby` or whatever), you can add other runners to `/etc/gitlab-runner/config.toml`. Please DO NOT overwrite `allowed_images = ["asclinux/linuxforphp-8.1-ultimate:*"]` for this runner, since it is a potential security breach.

Moreover, many services and languages are already available from within the **Linux for PHP** Docker images. Thus, by adding the appropriate `RUN` commands in the container's `Dockerfile`, one can launch all the needed services to quickly run a set of tests (Ruby, Python, Java, Node.js, Mongo server, Redis server, PostgreSQL server, etc.).

## Requirements
 - Gitlab v`9.0` and later
 - Gitlab runner v`9.0` and later

## Similar projects
 - https://github.com/bobey/docker-gitlab-ci-runner (For legacy gitlab-ci-runner)


CHANGELOG.md
# CHANGELOG - Linux for PHP Ultimate Docker images for GitLab CI

`2019-05-06`
 - Updates the repository to use Linux for PHP images
 - [PHP] Adds PHP 7.2 and 7.3
 - [PHP] Updates the PHP 5.6 and 7.0 modules
 
`2017-05-28`
 - Add yarn
 - memchache commented out in 7.0 and 7.1 due to compilation errors
 - XDebug updated to 2.5.4 for 7.0 and 7.1

`2017-04-08`
 - Clean up PHP < 5.5
 - Add igbinary

`2016-12-17`
 - [PHP] Added 7.1 image
 - [PHP] Xdebug updated 2.5.0 for PHP 5.5+

`2016-11-02`
 - [PHP] Update xdebug to 2.4.1 for PHP 5.4+
 - [PHP] Add rsync
 - [PHP] Fix compilation issues

`2016-06-20`
 - [PHP] Update xdebug to 2.4.0 for PHP 5.4+

`2016-06-14`
 - [PHP] Nodejs upgraded to 6.x

`2016-06-12`
 - [PHP] Add amqp extension
 - [PHP] Add memcache and redis extensions for PHP7

`2016-04-26`
 - [PHP] Uncomment enabling xdebug from environment variables for PHP7.0 (thanks to @vikpe)

`2016-03-27`
 - [PHP] Remove PHP 5.2 since Debian 6.0 image is not available on Docker hub anymore

`2016-03-13`
 - [PHP] Update PHP7 xdebug version to stable 2.4.0

`2016-03-04`
 - [VM] Disable updating docker images via cron since gitlab-runner is doing that itself

`2016-01-10`
 - [PHP] Add xdebug for PHP 7
 - [PHP] Disable xdebug by default for PHP 5.2

`2016-01-09`
 - [HHVM] Preliminary hhvm release

`2016-01-06`
- [PHP] *WARNING: Images with `-cli` tags are not supported anymore* to decrease docker hub building overhead. So use `tetraweb/php:5.6` instead of `tetraweb/php:5.6-cli`, they are equal
- [PHP] Add missing libicu library for `intl` extension
- [PHP] Add new `mongodb` to 5.4+ images including 7.0



php/7.3/entrypoint.sh
#!/bin/bash
set -e

# With env variable WITHOUT_XDEBUG=1 xdebug extension will be disabled
[ ! -z "$WITHOUT_XDEBUG" ] &&
export PHP_ZEND_EXTENSION_DIR = $( php -i | grep extensions | awk '{print $3}' ) &&
sed -i "s/zend_extension=${PHP_ZEND_EXTENSION_DIR//\//\\\/}\/xdebug.so/;zend_extension=${PHP_ZEND_EXTENSION_DIR//\//\\\/}\/xdebug.so/" /etc/php.ini

# Provide github token if you are using composer a lot in non-interactive mode
# Otherwise one day it will get stuck with request for authorization
# https://github.com/settings/tokens
if [[ ! -z "$COMPOSER_GITHUB" ]]
then
  composer config --global github-oauth.github.com "$COMPOSER_GITHUB"
fi

#
# If $TIMEZONE variable is passed to the image - it will set system timezone 
# and php.ini date.timezone value as well
# Overwise the default system Etc/UTC timezone will be used
#
# Also you can set the php timezone with direct setting it in php.ini 
# within your .gitlab-ci.yml like
# before_script:
# - sed -i "s/;date.timezone =/date.timezone = America\/New_York/" /etc/php.ini

if [[ ! -z "$TIMEZONE" ]]; then
  echo "$TIMEZONE" > /etc/timezone
  cp /usr/share/zoneinfo/$TIMEZONE /etc/localtime
  sed -i "s/;date.timezone =/date.timezone = ${TIMEZONE//\//\\\/}/" /etc/php.ini
fi

exec "$@"


gitlab-runner-vm/bootstrap.sh
#!/bin/bash

###############################################################################
#
# This is a bootstrap script for quickly setting up gitlab-runner
# on virtual (or even metal) machine with Ubuntu-14.04
#
# gitlab-runner is configured below to run the separate 
# docker container (runner) for each version of php
#
# TODO:
# System-related variables:
# USE_SWAP="1G" for creating swap
#
# Runner-related variables:
# CI_URL - URL of Gitlab-CI installation
# CI_TOKEN - Token for adding runner. Can be found ($CI_URL/admin/runners)
# CONCURRENT=1 Number of runners which can be executed simultaneously
# COMPOSER_GITHUB="TOKEN" - Github oauth token passed to each container
# TIMEZONE="America/New_York" for setting timezone
#
# Usage:
# 1. Install fresh ubuntu distribution
# 2.1 Non-interactive mode (doesn't require any input):
#     # export CI_URL=http://git.example.com/ci CI_TOKEN=12345abcdef[ OTHER_VAR=VALUE ...]; curl -SL https://raw.githubusercontent.com/TetraWeb/docker/master/gitlab-runner-vm/bootstrap.sh | bash
# OR
# 2.2 Interactive mode (script will ask for all variables to be typed while run)
#     # curl -SL https://raw.githubusercontent.com/linuxforphp/docker-gitlab-ci/master/gitlab-runner-vm/bootstrap.sh | bash
#
###############################################################################

set -e

create_swap() {
    dd if=/dev/zero of=/swapfile bs=1024 count=1024k
    mkswap /swapfile
    swapon /swapfile
    echo "/swapfile       none    swap    sw      0       0" >> /etc/fstab
    echo 10 | tee /proc/sys/vm/swappiness
    echo vm.swappiness = 10 | tee -a /etc/sysctl.conf
    chown root:root /swapfile
    chmod 0600 /swapfile
}

do_install() {
    while [ -z "$CI_URL" ]; do
        echo -n "Enter URL of your Gitlab CI installation [http://git.example.com/ci]: "
        read CI_URL < /dev/tty
    done

    while [ -z "$CI_TOKEN" ]; do
        echo -n "Enter API key (find it at $CI_URL/admin/runners): "
        read CI_TOKEN < /dev/tty
    done

    if [ -z "$TIMEZONE" ]; then
        echo -n "Timezone name [America/New_York]: "
        read TIMEZONE < /dev/tty
        if [ -z "$TIMEZONE" ]; then
            TIMEZONE="America/New_York"
        fi
        echo "$TIMEZONE" > /etc/timezone
        dpkg-reconfigure -f noninteractive tzdata
    fi

    if [ -z "$COMPOSER_GITHUB" ]; then
        echo -n "Composer Github token (optional): "
        read COMPOSER_GITHUB < /dev/tty
    fi

    if [ ! -z "$USE_SWAP" ]; then
        create_swap
    fi

    if [ -z "$CONCURRENT" ]; then
        echo -n "Number of concurrent processes ($(nproc)): "
        read CONCURRENT < /dev/tty
        if [ -z "$CONCURRENT" ]; then
            CONCURRENT=$(nproc)
        fi
    fi

    export DEBIAN_FRONTEND=noninteractive
    apt-get update && apt-get -y upgrade
    apt-get -y install mc htop ntpdate git curl wget openssh-server

    # Install docker if not found
    docker >/dev/null 2>&1 || { wget -qO- https://get.docker.com/ | sh; }

    # Install multi-runner
    curl -L https://packages.gitlab.com/install/repositories/runner/gitlab-ci-multi-runner/script.deb.sh | sudo bash
    apt-get install gitlab-ci-multi-runner

    # Declare environment variables array
    declare -a ENVVARS

    ENVVARS+=("MYSQL_ALLOW_EMPTY_PASSWORD=1")

    if [ ! -z "$COMPOSER_GITHUB" ]; then
        ENVVARS+=("COMPOSER_GITHUB=$COMPOSER_GITHUB")
    fi
    if [ ! -z "$TIMEZONE" ]; then
        ENVVARS+=("TIMEZONE=$TIMEZONE")
    fi

    gitlab-ci-multi-runner register -n -r "$CI_TOKEN" -u "$CI_URL" --tag-list 'php,mysql' --executor docker \
        --docker-image "asclinux/linuxforphp-8.1-ultimate:7.3-nts" --docker-allowed-images \
        "asclinux/linuxforphp-8.1-ultimate:*" --docker-allowed-services "*" --docker-allowed-services "*/*" \
        $(printf " --env %s" "${ENVVARS[@]}")

    sed -i -- "s/concurrent = 1/concurrent = $CONCURRENT/g" /etc/gitlab-runner/config.toml

    cronjob="#!/bin/bash\n"
#    Automatic updates of images seems to be working in gitlab-runner
#    for phpver in 5.6-nts 7.0-nts 7.1-nts 7.2-nts 7.3-nts
#    do
#        cronjob+="docker pull asclinux/linuxforphp-8.1-ultimate:$phpver\n"
#    done

    # Cleanup orphaned images
    cronjob+="docker rmi \$(docker images | grep none | awk '{print \$3}')\n"
    echo -e "$cronjob" > /etc/cron.weekly/docker-update-images
    chmod 755 /etc/cron.weekly/docker-update-images
}

do_install
